src/constraint.rs

```diff
@@ -1074,3 +1101,154 @@ impl ConstraintValidator for ElementConstraint {
         Ok(())
     }
 }
+
+/// Implements the `annotations` constraint
+/// [annotations]: https://amzn.github.io/ion-schema/docs/spec.html#annotations
+// The `required` annotation provided on the list of annotations is not represented here,
+// requirement of an annotation is represented in the annotation itself by the field `is_required` of `Annotation` struct.
+#[derive(Debug, Clone, PartialEq)]
+pub struct AnnotationsConstraint {
+    is_closed: bool,
+    is_ordered: bool,
+    annotations: Vec<Annotation>,
+}
+
+impl AnnotationsConstraint {
+    pub fn new(is_closed: bool, is_ordered: bool, annotations: Vec<Annotation>) -> Self {
+        Self {
+            is_closed,
+            is_ordered,
+            annotations,
+        }
+    }
+
+    pub fn validate_ordered_annotations(
+        &self,
+        value: &OwnedElement,
+        type_store: &TypeStore,
+        violations: Vec<Violation>,
+    ) -> ValidationResult {
+        let mut value_annotations = value
+            .annotations()
+            .map(|sym| sym.text().unwrap())
+            .peekable();
+
+        for expected_annotation in &self.annotations {
```

Comment:
I'm finding `validate_ordered_annotations` a bit difficult to wrap my head around. I wonder if it would be simpler to invert the iterators?:

```rust
for value_annotation in value_annotations {
  if value_annotation == expected_annotations.peek() {
    // We found an expected annotation! Move on to the next one.
    let _ = expected_annotations.next();
  } else if is_closed {
    // Uh oh, add it to the violations
  } else {
    // Open content
  }
}

// This iterator should be empty if we found everything we needed
while let Some(missing) = expected_annotations.next().is_some() {
  if missing.is_required() {
    // Violation
  }
}
```

It may also be that breaking things into some named helper functions would reduce the mental overhead involved.

Reply:
Added more comments and helper method.